### PR TITLE
updated ChangeLog file for 2.4.0

### PR DESCRIPTION
**Pull request by UnixJunkie**

### CODE REVIEW

**Comment by gasche** [comment]
Could you maybe remove the entries that correspond to internal changes only, with no visible difference for the user? For example I don't think anyone needs to know about "create testsuite-only targets" or "updated NEXT_RELEASE tags". I'll do line-by-line comments on stuff that I think could be skipped.

File: ChangeLog [hunk 1]
```diff
@@ -3,12 +3,106 @@ Changelog
...
 
-- Enum: uniq is modified to use structural comparison internally, instead
-    of previously used physical comparison, which was most likely a bug.
-    The old behavior is provided by the new Enum.uniqq function.
-    Enum.uniq_by was added for additional convenience.
+- ignore deprecation warnings
+  (Gabriel Scherer)
+- BatBitSet: use Bytes instead of String
+  (Gabriel Scherer)
+- rm qtest-generated file all_tests.ml upon 'make clean'
```
Comment (gasche) [inline ChangeLog]: skip

File: ChangeLog [hunk 2]
```diff
@@ -3,12 +3,106 @@ Changelog
 
 ## v2.4.0
 
-- Enum: uniq is modified to use structural comparison internally, instead
-    of previously used physical comparison, which was most likely a bug.
-    The old behavior is provided by the new Enum.uniqq function.
-    Enum.uniq_by was added for additional convenience.
+- ignore deprecation warnings
+  (Gabriel Scherer)
+- BatBitSet: use Bytes instead of String
```
Comment (gasche) [inline ChangeLog]: merge with the "deprecation warnings" and "corrected many warnings" entries (listing both authors), with some more general description like "update codebase wrt. new OCaml warnings -- preparation for future `-safe-string` compat".

File: ChangeLog [hunk 3]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Gabriel Scherer)
+- rm qtest-generated file all_tests.ml upon 'make clean'
+  (Francois Berenger)
+- corrected many warnings
+  (Francois Berenger)
+- regression tests for issue #609
```
Comment (gasche) [inline ChangeLog]: skip -- the fix entry "BatHashtbl: ..." is enough. When a fix was reported in the issue list, could you maybe add the PR number?

File: ChangeLog [hunk 4]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Gabriel Scherer)
+- Makefile: create testsuite-only-{byte,native} targets
```
Comment (gasche) [inline ChangeLog]: skip

File: ChangeLog [hunk 5]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Gabriel Scherer)
+- BatHashtbl: fix hash computation bug introduced by 4.01
+  (Gabriel Scherer)
+- BatText: synchronize nsplit to match BatString.nsplit
+  (Gabriel Scherer)
+- BatInt: repared quick check test for mid
+  (Francois Berenger)
+- updated NEXT_RELEASE tags
```
Comment (gasche) [inline ChangeLog]: skip

File: ChangeLog [hunk 6]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Francois Berenger)
+- BatLazyList: fix remove_all_such documentation
```
Comment (gasche) [inline ChangeLog]: PR number?

File: ChangeLog [hunk 7]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Xavier Van de Woestyne)
+- BatMap: better ocamldoc for update
```
Comment (gasche) [inline ChangeLog]: either skip or be more detailed "specify property X" -- as such, it does not help our users know what has changed or learn new thing about the library.

File: ChangeLog [hunk 8]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Francois Berenger)
+- BatMap: add pop_min_binding and pop_max_binding
+  (Francois Berenger)
+- BatMap: fix at_rank_exn compilation
```
Comment (gasche) [inline ChangeLog]: skip

File: ChangeLog [hunk 9]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Gabriel Scherer)
+- BatMap: modified update to behave as Set.update
```
Comment (gasche) [inline ChangeLog]: skip -- you already announce BatMap.update below

File: ChangeLog [hunk 10]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Francois Berenger)
+- BatMap: add at_rank_exn
+  (Francois Berenger)
+- BatSet: shorter exception strings in at_rank_exn
```
Comment (gasche) [inline ChangeLog]: skip

File: ChangeLog [hunk 11]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Francois Berenger)
+- BatMap: add update
+  (Francois Berenger)
+- BatEnum: add interleave utility function
```
Comment (gasche) [inline ChangeLog]: PR number?

File: ChangeLog [hunk 12]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Aleksey Z. Arens)
+- fix _tags warnings settings
```
Comment (gasche) [inline ChangeLog]: skip

File: ChangeLog [hunk 13]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Gabriel Scherer)
+- BatFormat: expose asprintf for V>=4.01
```
Comment (gasche) [inline ChangeLog]: PR number?

File: ChangeLog [hunk 14]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Ralf Vogler)
+- BatSet: more generic version of update
```
Comment (gasche) [inline ChangeLog]: what does this mean? What was the previous version, why is the new version more generic?

File: ChangeLog [hunk 15]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Francois Berenger)
+- BatSet: updated ocamldoc
```
Comment (gasche) [inline ChangeLog]: skip or more details: as an end-user, what do I learn from this entry?

File: ChangeLog [hunk 16]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Francois Berenger)
+- BatSet: add at_rank_exn
```
Comment (gasche) [inline ChangeLog]: maybe add the type to let people know how to use it

File: ChangeLog [hunk 17]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Francois Berenger)
+- BatUTF8: add qtests for enum
```
Comment (gasche) [inline ChangeLog]: skip; PR number for the entry below?

File: ChangeLog [hunk 18]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Kevin Brubeck Unhammer)
+- BatUTF8: add enum
+  (Kevin Brubeck Unhammer)
+- Move prefiltering logic from Makefile to ocamlbuild
```
Comment (gasche) [inline ChangeLog]: skip

File: ChangeLog [hunk 19]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Gabriel Scherer)
+- BatSet: add of_array
+  (Francois Berenger)
+- BatSet: add to_array and test for to_list
+  (Francois Berenger)
+- BatSet: add pop_max
+  (Francois Berenger)
+- BatSet: add pop_min
```
Comment (gasche) [inline ChangeLog]: maybe you could merge the pop_max and pop_min changes, and also of_array/to_array

File: ChangeLog [hunk 20]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Francois Berenger)
+- BatFormat: add new pp_* functions for old OCaml versions
```
Comment (gasche) [inline ChangeLog]: skip

File: ChangeLog [hunk 21]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Gabriel Scherer)
+- BatSplay: hardened against read-only data compiler optimization
+  (Gabriel Scherer)
+- BatList: simplified interleave
```
Comment (gasche) [inline ChangeLog]: what does "simplified" mean?

File: ChangeLog [hunk 22]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Francois Berenger)
+- BatOption: add Infix.(>>=)
+  (Herry Herry)
+- BatHashtbl: rearrange comment
```
Comment (gasche) [inline ChangeLog]: skip

File: ChangeLog [hunk 23]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Anders Fugmann)
+- BatHashtbl: add comments to modify_opt plus indentation changes
+  (Anders Fugmann)
+- BatHashtbl: modify now have the same semantic than replace
```
Comment (gasche) [inline ChangeLog]: same .. as ..

File: ChangeLog [hunk 24]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Anders Fugmann)
+- BatHashtbl: more efficient modify_opt and modify_def
+  (Anders Fugmann)
+- BatFormat: add comment
```
Comment (gasche) [inline ChangeLog]: skip or detail

File: ChangeLog [hunk 25]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Christoph Höger)
+- BatFormat: add pp_print_list and pp_print_text
+  (Christoph Höger)
+- LazyList: fix description of range
```
Comment (gasche) [inline ChangeLog]: skip or detail

File: ChangeLog [hunk 26]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Quentin Stievenart)
+- ChangeLog update
   (Max Mouratov)
-
+- add "@since 2.4.0" tag
+  (Max Mouratov)
+- BatEnum: add uniq_by and uniqq
+  (Max Mouratov)
+- BatEnum: fix uniq to use structural comparison
```
Comment (gasche) [inline ChangeLog]: maybe merge with the entry above?

File: ChangeLog [hunk 27]
```diff
@@ -3,12 +3,106 @@ Changelog
...
+  (Max Mouratov)
+- BatUnix: add run_and_read
+  (Francois Berenger)
+- BatCache: use hashmap to speed up lru_cache
+  (Sang Kil Cha)
+- add lexer location directives in prefilter.ml
```
Comment (gasche) [inline ChangeLog]: skip

**Comment by gasche** [comment]
The purpose of the Changelog, in my opinion, is not really to list development events (we have `git log` for that), but rather to let users know what has changed since the last version. A secondary (but also important) purpose it to credit contributors who worked on the release, in particular "external", that is infrequent, contributors (so I always leave their mention even if I would skip the entry if I had done it -- or you, as you are very much the main contributor of this release -- myself).

In particular, any entry should try to maximize the information given to the end-user in the amount of space. If you add a new function, it should be clear what it does from the name (hopefully most of our functions have this property), or else add a type signature or a short description. If you optimize a corner case, don't just write "optimize a corner case" but describe the case.

Finally, it is helpful if you can have the PR number in the entry, in case users want to go back to the change and its discussion. If this last part is too much work for you, it's acceptable to skip it.

**Comment by UnixJunkie** [comment]
I reduced it further based on your input.
I did not add the PR numbers; it was not done previously and having to update post mortem
the changelog for everybody's changes is painful and boring enough.
I feel, in the future, it should be updated incrementally with any PR by the person submitting the PR.
So that the next release can go a little faster.

File: ChangeLog [hunk 28]
```diff
@@ -3,12 +3,66 @@ Changelog
 
 ## v2.4.0
 
-- Enum: uniq is modified to use structural comparison internally, instead
-    of previously used physical comparison, which was most likely a bug.
-    The old behavior is provided by the new Enum.uniqq function.
-    Enum.uniq_by was added for additional convenience.
+- preparation for upcoming '-safe-string compat' OCaml warning
```
Comment (gasche) [inline ChangeLog]: I find this line hard to understand. I tried to propose a reformulation and failed to find something convincing. Could we just drop it?

**Comment by gasche** [comment]
Very nice, thanks! I had a minor comment but otherwise it's good for merging.

**Pull request merged**